### PR TITLE
Eth get uncle by block number and index

### PR DESCRIPTION
**Pull request by Sixtysixter**
Implementation of the API eth_getUncleByBlockNumberAndIndex

### CODE REVIEW

File: silkrpc/commands/eth_api.cpp [hunk 1]
```diff
@@ -296,12 +296,40 @@ asio::awaitable<void> EthereumRpcApi::handle_eth_get_uncle_by_block_hash_and_ind
...
+        auto idx = std::stoul(index);
+        if (idx >= ommers.size()) {
+            reply = make_json_content(request["id"], nullptr);
+            co_return;
+        }
+        const auto total_difficulty = co_await core::rawdb::read_total_difficulty(tx_database, block_with_hash.hash, block_number);
+        auto uncle = ommers[idx];
+
+        const silkworm::BlockBody block_body{};
+        const evmc::bytes32 hash {};
```
Comment (canepat) [inline silkrpc/commands/eth_api.cpp:328]: Lines 327 and 328 are unnecessary, see comment below

File: silkrpc/commands/eth_api.cpp [hunk 2]
```diff
@@ -296,12 +296,40 @@ asio::awaitable<void> EthereumRpcApi::handle_eth_get_uncle_by_block_hash_and_ind
...
+        silkworm::BlockWithHash block_hash{{block_body, uncle}, hash};
```
Comment (canepat) [inline silkrpc/commands/eth_api.cpp:329]: Some improvements here:
1) naming seems ambiguous (it's not an hash) and too similar with line 316
2) it should be more clear that we're creating an empty block to hold the uncle block header
3) `hash` is empty here, whilst it should be `uncle.hash()`

In short I suggest to replace line 329 with:

`silkworm::BlockWithHash uncle_block_with_hash{{{}, uncle}, uncle.hash()};`

File: silkrpc/commands/eth_api.cpp [hunk 3]
```diff
@@ -296,12 +296,40 @@ asio::awaitable<void> EthereumRpcApi::handle_eth_get_uncle_by_block_hash_and_ind
...
+        const Block new_block{block_hash, total_difficulty};
```
Comment (canepat) [inline silkrpc/commands/eth_api.cpp:330]: `new_block` -> `uncle_block_with_hash_and_td`

(`silkrpc::Block` should be renamed something like `silkrpc::BlockWithHashAndTD` but let's do this in a separate PR)

File: silkrpc/commands/eth_api.cpp [hunk 4]
```diff
@@ -296,12 +296,40 @@ asio::awaitable<void> EthereumRpcApi::handle_eth_get_uncle_by_block_hash_and_ind
...
 
     try {
         ethdb::TransactionDatabase tx_database{*tx};
 
-        reply = make_json_content(request["id"],  "0x" + to_hex_no_leading_zeros(0));
+        const auto block_number = co_await core::get_block_number(block_id, tx_database);
+        const auto block_with_hash = co_await core::rawdb::read_block_by_number(tx_database, block_number);
+        const auto ommers = block_with_hash.block.ommers;
+
+        auto idx = std::stoul(index);
```
Comment (canepat) [inline silkrpc/commands/eth_api.cpp:319]: Have you tried with `index` different from `0x0` (e.g. `0x1`)? This line doesn't work, it should be:

`auto idx = std::stoul(index, 0, 16);`

because `index` is in QUANTITY format (we should introduce a new `silkrpc::Quantity` type but we'll do in a separate PR)

**Review by canepat — CHANGES_REQUESTED** [review]

File: docs/API.md [hunk 5]
```diff
@@ -28,7 +28,7 @@ The following table shows the current [JSON RPC API](https://eth.wiki/json-rpc/A
 |                                         |              |                                            |
 | eth_getTransactionByHash                | -            | not yet implemented                        |
 | eth_getTransactionByBlockHashAndIndex   | -            | not yet implemented                        |
-| eth_getTransactionByBlockNumberAndIndex | -            | not yet implemented                        |
+| eth_getTransactionByBlockNumberAndIndex | Yes          |                                            |
```
Comment (canepat) [inline docs/API.md:31]: Please add in this PR also `eth_getUncleCountByBlockNumber` implementation because it's closely related to this one

Comment (Sixtysixter) [reply docs/API.md:31]: oopps.

File: silkrpc/commands/eth_api.cpp (hunk 1 above)
Comment (Sixtysixter) [reply silkrpc/commands/eth_api.cpp:328]: Done

**Review by canepat — APPROVED** [review]
